crates/miden-objects/src/note/note_tag.rs

```diff
@@ -41,29 +45,59 @@ pub enum NoteExecutionMode {
 ///   should be consumed by the network. These notes will be further validated and if possible
 ///   consumed by it.
-/// - Target describes how to further interpret the bits in the tag. For tags with a specific
-///   target, the rest of the tag is interpreted as a partial [`AccountId`]. For use case values,
-///   the meaning of the rest of the tag is not specified by the protocol and can be used by
-///   applications built on top of the chain.
+/// - Target describes how to further interpret the bits in the tag.
+///   - For tags with a specific target, the rest of the tag is interpreted as a partial
+///     [`AccountId`]. For network accounts these are the first 30 bits of the ID while for local
+///     account targets, the first 14 bits are used - a trade-off between privacy and uniqueness.
+///   - For use case values, the meaning of the rest of the tag is not specified by the protocol and
+///     can be used by applications built on top of the rollup.
 ///
 /// The note type is the only value enforced by the protocol. The rationale is that any note
 /// intended to be consumed by the network must be public to have all the details available. The
 /// public note for local execution is intended to allow users to search for notes that can be
 /// consumed right away, without requiring an off-band communication channel.
+///
+/// **Note on Type Safety**
+///
+/// Each enum variant contains the raw encoding of the note tag, where the first two bits
+/// _should_ correspond to the variant's prefix (as defined in the table above). However, because
+/// enum variants are always public, it is possible to instantiate this enum where this invariant
+/// does not hold, e.g. `NoteTag::NetworkAccount(0b11...)`. For that reason, the enum variants
+/// should take precedence in case of such a mismatch and the inner value **should not be accessed
+/// directly**. Instead, only rely on [`NoteTag::as_u32`] to access the encoded value, which will
+/// always return the correct value.
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
-pub struct NoteTag(u32);
+pub enum NoteTag {
+    /// Represents a tag for a note intended for network execution, targeted at a network account.
+    /// The note must be public.
+    NetworkAccount(u32),
+    /// Represents a tag for a note intended for network execution for a public use case. The note
+    /// must be public.
+    NetworkUseCase(u16, u16),
+    /// Represents a tag for a note intended for local execution for a public use case. The note
+    /// must be public.
+    LocalPublicAny(u32),
+    /// Represents a tag for a note intended for local execution.
+    ///
+    /// This is used for two purposes:
+    /// - A private use case note can be private/public/encrypted.
+    /// - A note targeted at any type of account.
+    ///
+    /// In all cases, the note can be of any type.
+    LocalAny(u32),
```

Comment:
```suggestion
    /// Represents a tag for a note intended for local execution.
    ///
    /// This is used for two purposes:
    /// - A public use case.
    /// - A note targeted at any type of account.
    ///
    /// In all cases, the note must be **public**.
    LocalPublicAny(u32),
    /// Represents a tag for a note intended for local execution.
    ///
    /// This is used for two purposes:
    /// - A private use case.
    /// - A note targeted at any type of account.
    ///
    /// In all cases, the note can be of any type.
    LocalAny(u32),
```

We can update the docs for `LocalPublicAny`.